### PR TITLE
Support for Group Replication Hostgroups

### PR DESCRIPTION
**Pull request by CyberLine**
#### Pull Request (PR) description
This PR adds the missing Support for proxysql group replication hostgroups.
It also fixes a Bug, that prevents this module from being working with P4 and agent 1.10 because of unknown Type definitions

### CODE REVIEW

**Comment by bastelfreak** [comment]
Hi @CyberLine, thanks for the PR. Can you please take a look at the inline comment I made? Please also check your github settings, your commits are signed with gpg (which is good) but github cannot verify it. I assume you haven't uploaded your public key.

File: manifests/init.pp [hunk 1]
```diff
@@ -203,11 +203,11 @@
 
   String $node_name = "${facts['fqdn']}:${admin_listen_port}",
   Boolean $manage_hostgroup_for_servers = $proxysql::params::manage_hostgroup_for_servers,
-  Optional[Proxysql::Server] $mysql_servers = undef,
-  Optional[Proxysql::User] $mysql_users = undef,
-  Optional[Proxysql::Hostgroup] $mysql_hostgroups = undef,
-  Optional[Proxysql::Rule] $mysql_rules = undef,
-  Optional[Proxysql::Scheduler] $schedulers = undef,
+  Optional[Tuple] $mysql_servers = undef,
```
[the remark this answers is not in the document]
Comment (CyberLine) [reply manifests/init.pp]: Yeah, but as mentioned, this does not work with Puppet 4.8 under Debian 9.5...
As you may know, working with puppet modules is a plenty of mess from dependency hell. But I think for this PR I can remove it and fill a Bug ticket instead.

**Comment by CyberLine** [comment]
I will check that. Can you take a look in the tests? They are failing with something strange SSH connection errors...

**Comment by mcrauwel** [comment]
that SSH connection error is just a warning... the container takes a little while longer to start so it needs 1 retry. The error is further down...

**Comment by CyberLine** [comment]
> that SSH connection error is just a warning... the container takes a little while longer to start so it needs 1 retry. The error is further down...

`An error occurred in an `after(:suite)` hook.
Failure/Error:
defstat(host, path)
ruby = ruby_command(host)
owner = on(host"#{ruby} -e 'require\"etc\"; puts (Etc.getpwuid(File.stat(\"#{path}\").uid).name)'").stdout.chomp
group = on(host"#{ruby} -e 'require\"etc\"; puts (Etc.getgrgid(File.stat(\"#{path}\").gid).name)'").stdout.chomp
mode  = on(host"#{ruby} -e 'puts (File.stat(\"#{path}\").mode & 0777).to_s(8)'").stdout.chomp.to_i
[owner, group, mode]
end
Excon::Error::Socket:
  wrong number of arguments (given 0, expected 2) (ArgumentError)`

whats that ...

**Comment by mcrauwel** [comment]
@CyberLine that looks like an issue in the build/test stack... I saw some chatter on #voxpupuli about this. I guess we are waiting for an upstream fix and then another modulesync update right @bastelfreak?

**Comment by bastelfreak** [comment]
this was an issue on beaker-puppet, it got fixed and released. I restarted CI.

**Comment by CyberLine** [comment]
I still have to add something from my test branch to my PR ...

**Comment by CyberLine** [comment]
`An error occurred while loading ./spec/acceptance/class_spec.rb.
Failure/Error:install_module_dependencies_on(hosts)
Beaker::Host::CommandFailure:
  Host 'ubuntu1404-64-1' exited with 1 running:
   puppet module install puppetlabs-apt -v 5.0.1
  Last 10 lines of output were:
Notice: Preparing to install into /etc/puppetlabs/code/environments/production/modules ...
Notice: Downloading from https://forgeapi.puppet.com ...
Error: Could not install module 'puppetlabs-apt' (???)
  	  No version of 'puppetlabs-apt' can satisfy all dependencies
  	    Use `puppet module install --ignore-dependencies` to install only this module`

**Comment by mcrauwel** [comment]
yeah I was reviewing the Readme file to in my [local branch](https://github.com/mcrauwel/puppet-proxysql/tree/readme-me-and-coc)

**Comment by CyberLine** [comment]
@mcrauwel i also added back Puppet4 Agent Tests here: https://travis-ci.org/CyberLine/puppet-proxysql/builds/428505011. While the Module supports Puppet 4, so i think the test should cover this too.

**Comment by mcrauwel** [comment]
@CyberLine yeah that is a valid point, however the default puppet is now 5 so you're test are still running on puppet5 
```
ubuntu1404-64-1 07:46:02$ apt-get install --force-yes  -y puppet-agent
  Reading package lists...
  Building dependency tree...
  Reading state information...
  The following NEW packages will be installed:
    puppet-agent
  0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
  Need to get 17.8 MB of archives.
  After this operation, 96.1 MB of additional disk space will be used.
  Get:1 http://apt.puppetlabs.com/ trusty/puppet puppet-agent amd64 5.5.6-1trusty [17.8 MB]
```

also we should consider that part as a separate PR

**Comment by bastelfreak** [comment]
This looks good at the moment. @CyberLine can you add an acceptance test please? I  think afterwards we can merge this.

----

@mcrauwel wasn't there a PR that added excessive acceptance tests? Or did I mix it up with another module....

File: examples/init.pp [hunk 2]
```diff
@@ -15,6 +15,11 @@
   mysql_hostgroups => [ { 'hostgroup 1' => { 'writer_hostgroup' => 1,
```
Comment (mcrauwel) [inline examples/init.pp]: tests require these => to be aligned :(

File: spec/acceptance/class_spec.rb [hunk 3]
```diff
@@ -54,6 +54,24 @@ class { 'proxysql':
         comment          => 'Test MySQL Cluster 10-30',
       }
 
+      proxy_mysql_group_replication_hostgroup { '5-2-10-11':
```
Comment (mcrauwel) [inline spec/acceptance/class_spec.rb]: there seems to be a problem with the name here...

Comment (CyberLine) [reply spec/acceptance/class_spec.rb]: Well, copy pasta error .. i did never before upstream PRs to puppet modules i use in Prod.

**Comment by mcrauwel** [comment]
thnx for your contribution @CyberLine

**Pull request merged**